foundation/org.eclipse.persistence.oracle/src/main/java/org/eclipse/persistence/platform/database/oracle/Oracle21Platform.java

```diff
@@ -0,0 +1,198 @@
+        } else {
+            super.setParameterValueInDatabaseCall(parameter, statement, index, session);
+        }
+    }
+
+    /**
+     * INTERNAL
+     * Set the parameter in the JDBC statement at the given index.
+     * This support a wide range of different parameter types, and is heavily optimized for common types.
+     * Handles Postgres specific PGobject instances.
+     *
+     * @param parameter the parameter to set
+     * @param statement target {@code CallableStatement} instance
+     * @param name name of the parameter in the statement
+     * @param session current database session
+     */
+    @Override
+    public void setParameterValueInDatabaseCall(
+            final Object parameter, final CallableStatement statement,
+            final String name, final AbstractSession session
+    ) throws SQLException {
+        if (parameter instanceof OracleJsonValue) {
+            statement.setObject(name, parameter, OracleType.JSON);
+        } else {
+            super.setParameterValueInDatabaseCall(parameter, statement, name, session);
+        }
+    }
+
+    // Oracle 21c specific JSON types support:
+    // Stores JsonValue instances as JSON.
+    /**
+     * INTERNAL:
+     * Convert JSON value field to JDBC statement type.
+     * Postgres JSON storage type is {@code JSONB} and target Java type is {@code PGobject}.
+     *
+     * @param <T> classification type
+     * @param jsonValue source JSON value field
+     * @return converted JDBC statement type
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T convertJsonValueToDataValue(final JsonValue jsonValue) throws PersistenceException {
+        if (jsonValue == null) {
+            return null;
+        }
+        final StringWriter sw = new StringWriter(128);
+        try (final JsonWriter jw = Json.createWriter(sw)) {
+            jw.write(jsonValue);
+        }
+        try (final Reader jr = new StringReader(sw.toString())) {
+            return (T) factory.createJsonTextValue(jr);
+        } catch (IOException e) {
+            throw new PersistenceException("Could not convert JsonValue to OracleJsonValue", e);
+        }
+    }
+
+    /**
+     * INTERNAL:
+     * Convert JDBC {@code ResultSet} type to JSON value field.
+     *
+     * @param jdbcValue source classification type value from JDBC
+     * @return converted JSON field value
+     */
+    public JsonValue convertDataValueToJsonValue(Object jdbcValue) {
+        if (jdbcValue == null) {
+            return null;
+        }
+        if (jdbcValue instanceof OracleJsonValue) {
+        // Depends on javax.json, not jakarta.json!
+        //    return ((OracleJsonValue) jdbcValue).wrap(JsonValue.class);
+            try (final JsonReader jr = Json.createReader(new StringReader(jdbcValue.toString()))) {
+                return jr.readValue();
+            }
+        }
+        throw new PersistenceException("Could not convert JDBC ResultSet type to JsonValue");
```

Comment:
What about resource bundle usage like e.g.
`throw new PersistenceException(ExceptionLocalization.buildMessage("json_cant_convert", null));`

Reply:
OK, I'm changing this. But I don't believe it makes sense in EE4J days. Did you ever saw any localization for ExceptionLocalizationResource?
Also messages bundle is in core, but it's being used everywhere (oracle extension in this case).